src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -0,0 +1,624 @@
+#include <Common/parseGlobs.h>
+
+namespace fs = std::filesystem;
+
+namespace ProfileEvents
+{
+    extern const Event EngineFileLikeReadFiles;
+}
+
+namespace CurrentMetrics
+{
+    extern const Metric StorageObjectStorageThreads;
+    extern const Metric StorageObjectStorageThreadsActive;
+    extern const Metric StorageObjectStorageThreadsScheduled;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPILE_REGEXP;
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+    extern const int FILE_DOESNT_EXIST;
+}
+
+StorageObjectStorageSource::StorageObjectStorageSource(
+    String name_,
+    ObjectStoragePtr object_storage_,
+    ConfigurationPtr configuration_,
+    const ReadFromFormatInfo & info,
+    std::optional<FormatSettings> format_settings_,
+    ContextPtr context_,
+    UInt64 max_block_size_,
+    std::shared_ptr<IIterator> file_iterator_,
+    bool need_only_count_)
+    : SourceWithKeyCondition(info.source_header, false)
+    , WithContext(context_)
+    , name(std::move(name_))
+    , object_storage(object_storage_)
+    , configuration(configuration_)
+    , format_settings(format_settings_)
+    , max_block_size(max_block_size_)
+    , need_only_count(need_only_count_)
+    , read_from_format_info(info)
+    , create_reader_pool(std::make_shared<ThreadPool>(
+        CurrentMetrics::StorageObjectStorageThreads,
+        CurrentMetrics::StorageObjectStorageThreadsActive,
+        CurrentMetrics::StorageObjectStorageThreadsScheduled,
+        1/* max_threads */))
+    , columns_desc(info.columns_description)
+    , file_iterator(file_iterator_)
+    , schema_cache(StorageObjectStorage::getSchemaCache(context_, configuration->getTypeName()))
+    , create_reader_scheduler(threadPoolCallbackRunnerUnsafe<ReaderHolder>(*create_reader_pool, "Reader"))
+{
+}
+
+StorageObjectStorageSource::~StorageObjectStorageSource()
+{
+    create_reader_pool->wait();
+}
+
+void StorageObjectStorageSource::setKeyCondition(const ActionsDAGPtr & filter_actions_dag, ContextPtr context_)
+{
+    setKeyConditionImpl(filter_actions_dag, context_, read_from_format_info.format_header);
+}
+
+std::shared_ptr<StorageObjectStorageSource::IIterator> StorageObjectStorageSource::createFileIterator(
+    ConfigurationPtr configuration,
+    ObjectStoragePtr object_storage,
+    bool distributed_processing,
+    const ContextPtr & local_context,
+    const ActionsDAG::Node * predicate,
+    const NamesAndTypesList & virtual_columns,
+    ObjectInfos * read_keys,
+    std::function<void(FileProgress)> file_progress_callback)
+{
+    if (distributed_processing)
+        return std::make_shared<ReadTaskIterator>(
+            local_context->getReadTaskCallback(),
+            local_context->getSettingsRef().max_threads);
+
+    if (configuration->isNamespaceWithGlobs())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expression can not have wildcards inside namespace name");
```

Comment:
Maybe we can improve exception message for S3 and change "namespace" to "bucket", I think "namespace" can be confusing